tls/s2n_psk.c

```diff
@@ -278,56 +331,114 @@ S2N_RESULT s2n_finish_psk_extension(struct s2n_connection *conn)
         struct s2n_psk *psk = NULL;
-        GUARD_AS_POSIX(s2n_array_get(&conn->psk_params.psk_list, i_index, (void**) &psk));
-        ENSURE_POSIX_REF(psk);
+        GUARD_RESULT(s2n_array_get(&params->psk_list, i - 1, (void**)&psk));
+        ENSURE_REF(psk);
         if (psk->type == S2N_PSK_TYPE_EXTERNAL) {
-            GUARD(s2n_psk_free(psk));
-            GUARD_AS_POSIX(s2n_array_remove(&conn->psk_params.psk_list, i_index));
+            continue;
+        }
+        GUARD_RESULT(s2n_array_remove(&params->psk_list, i - 1));
+    }
+
+    /* Wipe only the external PSKs and the list itself */
+    GUARD_RESULT(s2n_psk_parameters_wipe(params));
+    return S2N_RESULT_OK;
+}
+
+int s2n_connection_set_external_psks(struct s2n_connection *conn, struct s2n_psk **psk_list, uint32_t psk_list_len)
+{
+    notnull_check(conn);
+    
+    /* To avoid corrupting existing PSKs on a failure, we will operate on a copy until successful */
+    DEFER_CLEANUP(struct s2n_psk_parameters new_params = { 0 }, s2n_cleanup_after_set_external_psks);
+    GUARD_AS_POSIX(s2n_psk_parameters_init(&new_params));
+
+    /* Copy any non-external PSKs into the new list */
+    for (uint32_t i = 0; i < conn->psk_params.psk_list.len; i++) {
+        struct s2n_psk *existing_psk = NULL;
+        GUARD_AS_POSIX(s2n_array_get(&conn->psk_params.psk_list, i, (void**) &existing_psk));
+        notnull_check(existing_psk);
+
+        if (existing_psk->type == S2N_PSK_TYPE_EXTERNAL) {
+            continue;
         }
+        GUARD_AS_POSIX(s2n_array_insert_and_copy(&new_params.psk_list, new_params.psk_list.len, existing_psk));
     }
 
-    for (size_t i = 0; i < psk_vec_length; i++) {
+    if (psk_list_len != 0) {
+        notnull_check(psk_list);
+    }
+
+    for (uint32_t i = 0; i < psk_list_len; i++) {
+        notnull_check(psk_list[i]);
+
         /* Check for duplicate identities */
-        size_t array_len = conn->psk_params.psk_list.len;
-        for (size_t j = 0; j < array_len; j++) {
-            struct s2n_psk *psk = NULL;
-            GUARD_AS_POSIX(s2n_array_get(&conn->psk_params.psk_list, j, (void**) &psk));
-            ENSURE_POSIX_REF(psk);
-            if (psk->identity.size == psk_vec[i].identity_length) {
-                ENSURE_POSIX(memcmp(psk->identity.data, psk_vec[i].identity, psk->identity.size) != 0, S2N_ERR_DUPLICATE_PSK_IDENTITIES);
-            }
+        uint32_t array_len = new_params.psk_list.len;
+        for (uint32_t j = 0; j < array_len; j++) {
+            struct s2n_psk *existing_psk = NULL;
+            GUARD_AS_POSIX(s2n_array_get(&new_params.psk_list, j, (void**) &existing_psk));
+            notnull_check(existing_psk);
+
+            bool duplicate = existing_psk->identity.size == psk_list[i]->identity.size
+                    && memcmp(existing_psk->identity.data, psk_list[i]->identity.data, existing_psk->identity.size) == 0;
+            ENSURE_POSIX(!duplicate, S2N_ERR_DUPLICATE_PSK_IDENTITIES);
```

Comment:
Since we are not including any of the external psks in the new_params list, how would we obtain duplicates here for external psks. Does this mean this logic will never be true except for resumption psks - for which this function is not relevant?

Reply:
We can't allow external PSKs with identities that match resumption PSKs. No PSKs, regardless of type, should have the same identity.

But probably more importantly, the duplicate check is performed in the same loop that's adding external PSKs to the list. That means we're also comparing a new external PSK to every other new PSK already added to the list, ensuring there also aren't any duplicates in the input list.